### DISCUSSION TITLE
Typescript and select() queries

### DISCUSSION THREAD

**Discussion by jonlambert** [supabase/supabase#2465] [Questions] [answered]
Hey everyone,

As [explained in the docs](https://supabase.io/docs/reference/javascript/select), when running a select() query we can define fields to return as a comma-separated list. How do you go about retaining types with a query like this?

```typescript
const { data, error } = await supabase
  .from('users')
  .select(`
    name,
    cities (
      name
    )
  `)
```

Is it a case of manually building the response type by `Pick<>`ing etc? I'd need to cast the result to `any` in order to support this so it doesn't quite feel right.

Would be great to hear how you all approach this problem. It might be obvious but I can't see a solution in the docs. Thanks in advance!

**Comment by silentworks** [comment]
I myself don't use TypeScript with the supabase-js library, but I would like to think you would just create an interface that is modelled around the type of data you would get back from the `users` table.

**Comment by dshukertjr** [comment, marked as answer]
Hi @jonlambert 

Expanding on @silentworks's post, you can cast types to your data like this:

```typescript
  interface User {
    name: string
    cities: { name: string }[]
  }

  const { data, error } = await supabase.from<User>(
    'users'
  ).select(`
    name,
    cities (
      name
    )
  `)
```

Or you can directly cast the type like this:

```typescript
  const { data, error } = await supabase.from<{ name: string; cities: { name: string }[] }>(
    'users'
  ).select(`
    name,
    cities (
      name
    )
  `)
```

**Reply by jonlambert** [reply]
Makes sense - thanks @silentworks and @dshukertjr. Was hoping to be able to still take advantage of the generated tables, but guess I can just do that manually using Pick<> and Omit<>. Looks like I'll have to do the same with JSON columns anyway!
